### ISSUE TITLE
Rust-gdb print different integer value after add optimization flag

### ISSUE THREAD

**Issue by yuanboli233** [rust-lang/rust#59246]
The issue is related to the #59149, but I think this issue post presents the problem much clearer.

The rust-gdb debugger give incorrect value for optimization level 1 for the following "main.rs".
```
ybli@xxx:$ cat main.rs 
fn tmp() {
    println!(" ");    
}
fn main() {
    let mut counter = 0;
    {
        let mut tick = || counter += 1;
        tick();
        tick();
        tmp();
    }
    assert_eq!(counter, 2);
}
```

Rust-gdb will give different value for variable "counter" at the line "tmp();". For opt-level=0, it shows the variable has value 2; for opt-level=1, it shows the variable has value 0. Noticing that the program does not complain in the assert statement later, it seems that the program pass the wrong debugging information to the debugger.

opt-level=0:
```
ybli@xxx: $ rustc -g -C opt-level=0 main.rs 
ybli@xxx: $ rust-gdb ./main
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
[...... (omit some gdb description here)]
Reading symbols from ./main...done.
(gdb) b main.rs:10
Breakpoint 1 at 0x610d: file main.rs, line 10.
(gdb) r
Starting program: /home/ybli/Projects/tmp/experiment/main 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, main::main::hfe98083a4c87500f () at main.rs:10
10	        tmp();
(gdb) p counter
$1 = 2
```

opt-level=1:
```
ybli@xxx:$ rustc -g -C opt-level=1 main.rs 
ybli@xxx:$ rust-gdb ./main
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
[...... (omit some gdb description here)]
Reading symbols from ./main...done.
(gdb) b main.rs:10
Breakpoint 1 at 0x6050: file main.rs, line 10.
(gdb) r
Starting program: /home/ybli/Projects/tmp/experiment/main 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, main::main::hfe98083a4c87500f () at main.rs:10
10	        tmp();
(gdb) p counter
$1 = 0
```

## Meta
```
ybli@xxx:$ rustc --version --verbose
rustc 1.35.0-nightly (9d71ec135 2019-03-10)
binary: rustc
commit-hash: 9d71ec1358ac063fe6ff1eaed0ba6ed3cedde610
commit-date: 2019-03-10
host: x86_64-unknown-linux-gnu
release: 1.35.0-nightly
LLVM version: 8.0
```

**Comment by michaelwoerister** [comment]
Thanks for the bug report, @yuanboli233. This does indeed look like a bug, probably due to LLVM not preserving debuginfo across optimizations or optimizations shuffling around statements, making the breakpoint being set in an unexpected location.

**Comment by pnkfelix** [comment]
Visiting for wg-debugging triage

1. We should have someone go and confirm that it still reproduces.
2. If it *does* still reproduce, then leave E-help-wanted on, but now the E-help wanted will become a request for someone to help turn the issue into something that we can file on the LLVM bug repository.

I'm adding E-help-wanted to reflect the two steps above.

@rustbot label: E-help-wanted

**Comment by hexagonal-sun** [comment]
It looks like this may now be fixed:

```shell
/tmp ❯ rustc -g -C opt-level=1 main.rs
/tmp ❯ rust-gdb ./main
(gdb) b main.rs:10
Breakpoint 1 at 0x883a: file main.rs, line 12.
(gdb) r
Starting program: /tmp/main
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/libthread_db.so.1".


Breakpoint 1, main::main () at main.rs:12
12	                assert_eq!(counter, 2);
(gdb) p counter
$1 = <optimized out>
```

Notice that gdb prints that the value has been optimized away.

**Comment by Enselic** [comment]
Triage: Removing E-help-wanted since step 1 in https://github.com/rust-lang/rust/issues/59246#issuecomment-1251145158 has been performed, and step 2 is thus not needed.

I doubt adding E-needs-test will help for this issue, so I'll go ahead and close it. The referenced issue has also been closed ([motivation](https://github.com/rust-lang/rust/issues/59149#issuecomment-1597331356)).
